runtime/compiler/env/J9ClassEnv.cpp

```diff
@@ -417,24 +418,81 @@ J9::ClassEnv::isAnonymousClass(TR::Compilation *comp, TR_OpaqueClassBlock *clazz
    return comp->fej9()->isAnonymousClass(clazz);
    }
 
-const TR::TypeLayout*
-J9::ClassEnv::enumerateFields(TR::Region& region, TR_OpaqueClassBlock * opaqueClazz, TR::Compilation *comp)
-   {  
-   J9Class *clazz = (J9Class*)opaqueClazz;
-   TR_VMFieldsInfo fieldsInfo(comp, clazz, 1, stackAlloc);
-   ListIterator<TR_VMField> iter(fieldsInfo.getFields());
-   TR::TypeLayoutBuilder tlb(region);
-   for (TR_VMField *field = iter.getFirst(); field; field = iter.getNext())
+/*
+ * \param prefix
+ *    prefix is ended with `.`
+ */
+static char * mergeFieldNames(char * prefix, char * fieldName, bool withDotAtTheEnd, TR::Region &region)
+   {
+   int prefixLength = prefix ? strlen(prefix) : 0;
+   int nameLength = strlen(fieldName);
+   int mergedLength = nameLength + prefixLength;
+
+   if (withDotAtTheEnd)
+      mergedLength++;
+   mergedLength++; /* for adding \0 at the end */;
+
+   char * newName = new (region) char[mergedLength];
+   if (prefixLength > 0)
+      strncpy(newName, prefix, prefixLength);
+
+   strncpy(newName + prefixLength, fieldName, nameLength);
+
+   if (withDotAtTheEnd)
+      newName[mergedLength-2] = '.';
+   newName[mergedLength-1] = '\0';
+   return newName;
+   }
+
+static bool isFieldShapeQType(J9ROMFieldShape * shape)
+   {
+   // QType field signature example: "QNestedB;"
+   bool isQType = false;
+   J9UTF8 * signature = J9ROMNAMEANDSIGNATURE_SIGNATURE(&shape->nameAndSignature);
+   U_8 * classNameData = J9UTF8_DATA(signature);
+   U_16 classNameLength = J9UTF8_LENGTH(signature);
+
+   if ((classNameLength > 0) &&
+       (';' == classNameData[classNameLength - 1]) &&
+       ('Q' == classNameData[0]))
+      {
+      isQType = true;
+      }
+   return isQType;
+   }
+
+static void addEntryForField(TR_VMField *field, TR::TypeLayoutBuilder &tlb, TR::Region& region, J9Class * definingClass,
+                             char * prefix, IDATA offsetBase, TR::Compilation * comp)
+   {
+   J9JavaVM *vm = comp->fej9()->getJ9JITConfig()->javaVM;
+   if (isFieldShapeQType(field->shape) &&
+       vm->internalVMFunctions->isFlattenableFieldFlattened(definingClass, field->shape))
```

Comment:
Leo @Leonardo2718 is working on using helpers to load and store flattened fields.  Will those changes need symrefs for the aggregate fields to exist?  I'm just wondering whether the flattening of the symrefs should happen unconditionally like this or whether they should not be flattened if the JIT will use the load and store helpers.

Reply:
The code I have doesn't need the symrefs for the aggregate fields. The helpers can handle both flattened and unflattened fields so my code just needs to check whether the field is a QType. The new symrefs will be needed however if we want to add fastpaths around the calls (or at least have some way of determining whether a flattenable if is flattened).